tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -4464,32 +4603,56 @@ port::Status CreateOpRunners(
   };
 
   if (!use_fallback) {
-    auto heuristics = cudnn_frontend::EngineHeuristicsBuilder()
-                          .setOperationGraph(*op_graph)
-                          .setHeurMode(GetCudnnFrontendHeurMode())
-                          .build();
-    RETURN_MSG_IF_CUDNN_ERROR(heuristics);
+    std::optional<cudnn_frontend::EngineHeuristics> heuristics;
+    auto mode_b = cudnn_frontend::EngineHeuristicsBuilder()
+                      .setOperationGraph(*op_graph)
+                      .setHeurMode(GetCudnnFrontendHeurMode())
+                      .build();
+    if (mode_b.get_status() != CUDNN_STATUS_SUCCESS) {
+      // The mode_b can fallback to mode_instant if it cannot support things.
+      // However, there is a known issue in cudnn when dealing with runtime
+      // fusion engines. So, we manually do that here.
+      // TODO(kaixih@nvidia): remove this when the cudnn fixes it.
+      heuristics.emplace(cudnn_frontend::EngineHeuristicsBuilder()
+                             .setOperationGraph(*op_graph)
+                             .setHeurMode(CUDNN_HEUR_MODE_INSTANT)
+                             .build());
+      RETURN_MSG_IF_CUDNN_ERROR(*heuristics);
+    } else {
+      heuristics.emplace(std::move(mode_b));
+    }
 
     // cuDNN frontend sneakily puts error messages on the object and returns
     // partially-initialized results when there's an error; make sure to check
     // them.
-    int64_t engine_count = heuristics.getEngineConfigCount();
-    RETURN_MSG_IF_CUDNN_ERROR(heuristics);
-    auto& heuristics_configs = heuristics.getEngineConfig(engine_count);
-    RETURN_MSG_IF_CUDNN_ERROR(heuristics);
+    int64_t engine_count = heuristics->getEngineConfigCount();
+    RETURN_MSG_IF_CUDNN_ERROR(*heuristics);
+    auto& heuristics_configs = heuristics->getEngineConfig(engine_count);
+    RETURN_MSG_IF_CUDNN_ERROR(*heuristics);
```

Comment:
Aside: What you have done with using an optional to build different flavors of a move-only `EngineHeuristics` object is totally fine here, so no change necessary. However, for the sake of depicting another way to achieve this, the other option would have been to use a self-executing lambda as below. The minor advantages would have been not using optional and having to change the `.` to `->`.

```
auto heuristics = [&]() {
     auto mode_b = cudnn_frontend::EngineHeuristicsBuilder()
                       .setOperationGraph(*op_graph)
                       .setHeurMode(GetCudnnFrontendHeurMode())
                       .build();
     if (mode_b.get_status() != CUDNN_STATUS_SUCCESS) {
       // The mode_b can fallback to mode_instant if it cannot support things.
       // However, there is a known issue in cudnn when dealing with runtime
       // fusion engines. So, we manually do that here.
       // TODO(kaixih@nvidia): remove this when the cudnn fixes it.
       return cudnn_frontend::EngineHeuristicsBuilder()
                              .setOperationGraph(*op_graph)
                              .setHeurMode(CUDNN_HEUR_MODE_INSTANT)
                              .build();
     } else {
       return mode_b;
     }
}();
RETURN_MSG_IF_CUDNN_ERROR(heuristics);
```

Reply:
I am not too worried about this part. The cudnn has provided a new utility function to query the heuristics and it can do this fallback WAR. I am going to change this part when it is released.